itest/lnd_multi-hop_test.go

```diff
@@ -1838,3 +1838,165 @@ func createThreeHopNetwork(ht *lntest.HarnessTest,
 
 	return aliceChanPoint, bobChanPoint, carol
 }
+
+// testExtraMempoolPreimage tests that in the multi-hop setting,
+// Alice->Bob->Carol, if the outgoing HTLC is about to time out, Alice will go
+// on chain to claim it. Meanwhile, Bob decides to spend the HTLC on-chain
+// using Alice's commitment transaction via the preimage path. In this
+// scenario, the node that sent the outgoing HTLC(Alice) should extract the
+// preimage from the sweep transaction as soon as the transaction appears in
+// mempool, and finish settling the HTLC backwards into the route. In addition,
+// Alice should also try to claim the on-chain HTLC by fee bumping the 2nd
+// level timeout transaction.
+//
+// TODO(yy): update the test to check fee bumping is in effect.
+func testExtractMempoolPreimage(ht *lntest.HarnessTest) {
+	runMultiHopHtlcClaimTest(ht, runExtraPreimageFromMempool)
+}
+
+func runExtraPreimageFromMempool(ht *lntest.HarnessTest,
+	alice, bob *node.HarnessNode, c lnrpc.CommitmentType, zeroConf bool) {
+
+	// First, we'll create a three hop network: Alice -> Bob -> Carol, with
+	// Carol refusing to actually settle or directly cancel any HTLC's
+	// self.
+	aliceChanPoint, bobChanPoint, carol := createThreeHopNetwork(
+		ht, alice, bob, false, c, zeroConf,
+	)
+
+	// With the network active, we'll now add a new hodl invoice at Carol's
+	// end. Make sure the cltv expiry delta is large enough, otherwise Bob
+	// won't send out the outgoing htlc.
+	const invoiceAmt = 100000
+	var preimage lntypes.Preimage
+	copy(preimage[:], ht.Random32Bytes())
+	payHash := preimage.Hash()
+	invoiceReq := &invoicesrpc.AddHoldInvoiceRequest{
+		Value:      invoiceAmt,
+		CltvExpiry: finalCltvDelta,
+		Hash:       payHash[:],
+	}
+	eveInvoice := carol.RPC.AddHoldInvoice(invoiceReq)
+
+	// Subscribe the invoice.
+	stream := carol.RPC.SubscribeSingleInvoice(payHash[:])
+
+	// Now that we've created the invoice, we'll send a single payment from
+	// Alice to Carol. We won't wait for the response however, as Carol
+	// will not immediately settle the payment.
+	req := &routerrpc.SendPaymentRequest{
+		PaymentRequest: eveInvoice.PaymentRequest,
+		TimeoutSeconds: 60,
+		FeeLimitMsat:   noFeeLimitMsat,
+	}
+	alice.RPC.SendPayment(req)
+
+	// At this point, all 3 nodes should now have an active channel with
+	// the created HTLC pending on all of them.
+	htlc := ht.AssertHLTCActive(alice, aliceChanPoint, payHash[:])
+	ht.AssertHLTCActive(bob, bobChanPoint, payHash[:])
+	ht.AssertHLTCActive(carol, bobChanPoint, payHash[:])
+
+	// We let Alice go offline so the link between Alice and Bob is broken.
+	restartAlice := ht.SuspendNode(alice)
+
+	// Wait for carol to mark invoice as accepted. There is a small gap to
+	// bridge between adding the htlc to the channel and executing the exit
+	// hop logic.
+	ht.AssertInvoiceState(stream, lnrpc.Invoice_ACCEPTED)
+
+	// Carol now settles the invoice, which means Bob now has the preimage.
+	carol.RPC.SettleInvoice(preimage[:])
```

Comment:
Shouldn't we instead have Carol _force close_ to cause here to go on chain? Or rather, mine enough blocks for things to expire (taking Bob offline before hand), and only restart Bob once Carol tries to sweep with the second level preimage?